### PR TITLE
More docs for schedule api

### PR DESCRIPTION
**Pull request by duke8253**

### CODE REVIEW

**Comment by duke8253** [comment]
[approve ci Ubuntu]

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst [hunk 1]
```diff
@@ -55,6 +55,69 @@ called and continuations that use them have the same restrictions. ``TS_THREAD_P
 are threads that exist to perform long or blocking actions, although sufficiently long operation can
 impact system performance by blocking other continuations on the threads.
 
+Example Scenarios
+=================
+
+Scenario 1 (no thread affinity info, different types of threads)
+----------------------------------------------------------------
+
+NO affine thread set, plugin is calling the API on thread "A", "A" is of type "ET_TASK", and plugin
```
Comment (mlibbey) [inline doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: How about

When thread affinity is not set, a plugin calls the API on thread "A" (which is an "ET_TASK" type), and wants to schedule on an "ET_NET" type thread provided in "tp", the system would see there is no thread affinity information stored in "contp."

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst [hunk 2]
```diff
@@ -55,6 +55,69 @@ called and continuations that use them have the same restrictions. ``TS_THREAD_P
...
+wants to schedule on a "ET_NET" type thread provided in "tp".
+
+In this situation, system sees there is no thread affinity information stored in "contp". It then
+checks whether the type of thread "A" is the same as provided in "tp", and sees that "A" is "ET_TASK",
+but "tp" says "ET_NET". So "contp" gets scheduled on the next available "ET_NET" thread provided by a
+round robin list, which we will call thread "B". Since "contp" doesn't have thread affinity information,
+thread "B" will be assigned as the affine thread for it automatically.
+
+The reason for doing this is most of the time people want to schedule the same things on the same type
+of thread, so logically it is better to default the first thread that it is scheduled on as the affine
```
Comment (mlibbey) [inline doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: affine --> affinity?

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst [hunk 3]
```diff
@@ -55,6 +55,69 @@ called and continuations that use them have the same restrictions. ``TS_THREAD_P
...
+=================
+
+Scenario 1 (no thread affinity info, different types of threads)
+----------------------------------------------------------------
+
+NO affine thread set, plugin is calling the API on thread "A", "A" is of type "ET_TASK", and plugin
+wants to schedule on a "ET_NET" type thread provided in "tp".
+
+In this situation, system sees there is no thread affinity information stored in "contp". It then
+checks whether the type of thread "A" is the same as provided in "tp", and sees that "A" is "ET_TASK",
```
Comment (mlibbey) [inline doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: "tp," ..."ET_TASK," <== commas inside the quote. Same for next line's ET_NET's period (and "B."....)

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst [hunk 4]
```diff
@@ -55,6 +55,69 @@ called and continuations that use them have the same restrictions. ``TS_THREAD_P
...
+thread.
+
+Scenario 2 (no thread affinity info, same types of threads)
+-----------------------------------------------------------
+
+Slight variation of scenario 1, instead of scheduling on a "ET_NET" thread, the plugin wants to schedule
+on a "ET_TASK" thread (i.e. "tp" contains "ET_TASK" now), all other conditions stays the same.
+
+This time since the type of the desired thread for scheduling and thread "A" are the same, the system
+schedules "contp" on thread "A", and assigns thread "A" as the affine thread for "contp".
```
Comment (mlibbey) [inline doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: "A," <==comma inside ... "contp." <== period inside

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst [hunk 5]
```diff
@@ -55,6 +55,69 @@ called and continuations that use them have the same restrictions. ``TS_THREAD_P
...
+Scenario 3 (has thread affinity info, different types of threads)
+-----------------------------------------------------------------
+
+Slight variation of scenario 1, thread affinity is set for continuation "contp" to thread "A", all other
+conditions stays the same.
+
+In this situation, the system sees that the "tp" has "ET_NET", but the type of thread "A" is "ET_TASK".
+So even though "contp" has an affine thread, the system will not use that information since the type is
+different, instead it schedules "contp" on the next available "ET_NET" thread provided by a round robin
+list, which we will call thread "B". The difference with scenario 1 is that since thread "A" is set to
```
Comment (mlibbey) [inline doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: "B."

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst [hunk 6]
```diff
@@ -55,6 +55,69 @@ called and continuations that use them have the same restrictions. ``TS_THREAD_P
...
+be the affine thread for "contp" already, the system will NOT overwrite that information with thread "B".
```
Comment (mlibbey) [inline doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: affine ==> affinity?
"B."

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst [hunk 7]
```diff
@@ -55,6 +55,69 @@ called and continuations that use them have the same restrictions. ``TS_THREAD_P
...
+
+Most of the time, a continuation will be scheduled on one type of threads, and rarely gets scheduled on
+a different type. But when that happens we want it to return to the thread it was previously on, so it
```
Comment (mlibbey) [inline doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: But when that happens, <== add comma

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst [hunk 8]
```diff
@@ -55,6 +55,69 @@ called and continuations that use them have the same restrictions. ``TS_THREAD_P
...
+won’t have any lock contention problems. And that’s also why "thread_affinity" is not a hashmap of thread
+types and thread pointers.
+
+Scenario 4 (has thread affinity info, same types of threads)
+------------------------------------------------------------
+
+Slight variation of scenario 3, the only difference is "tp" now says "ET_TASK".
+
+This is the easiest scenario since the type of thread "A" and "tp" are the same, so the system schedules
+"contp" on thread "A". And as discussed before, there is really no reason why one may want to schedule
```
Comment (mlibbey) [inline doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: "A." And, as discussed, ...[add comma, remove 'before']

File: doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst (hunk 3 above)
Comment (duke8253) [reply doc/developer-guide/api/functions/TSContScheduleOnPool.en.rst]: Don't think this matters, just different styles, and they're not real quotes, just emphasize it's referencing the variable name.

**Comment by duke8253** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
Looks good.

**Pull request merged**